audit/audit.go

```diff
@@ -0,0 +1,130 @@
+package audit
+
+//this can't import kafka, as kafka already imports log
+//this either has to go in kafka or a separate package
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/ONSdigital/go-ns/identity"
+	"github.com/ONSdigital/go-ns/log"
+	"net/http"
+	"time"
+)
+
+type contextKey string
+
+const eventContextKey = contextKey("audit")
+const fromContextKey = contextKey("from")
+const requestIDHeader = "X-Request-Id"
+const fromHeader = "from"
+
+type Event struct {
+	Created         string         `avro:"created"`
+	Service         string         `avro:"service"`
+	Namespace       string         `avro:"name_space"`
+	RequestID       string         `avro:"request_id"`
+	User            string         `avro:"user"`
+	AttemptedAction string         `avro:"attempted_action"`
+	Result          string         `avro:"result"`
+	Params          []keyValuePair `avro:"params"`
+}
+
+type keyValuePair struct {
+	Key   string `avro:"key"`
+	Value string `avro:"value"`
+}
+
+type OutboundProducer interface {
+	Output() chan []byte
+}
+
+type Auditor struct {
+	Namespace string
+	TokenName string
+	Producer  OutboundProducer
+}
+
+func New(namespace string, producer OutboundProducer, token string) *Auditor {
+	return &Auditor{
+		Namespace: namespace,
+		Producer:  producer,
+		TokenName: token,
+	}
+}
+
+func (a *Auditor) Interceptor() func(handler http.Handler) http.Handler {
+	return func(handler http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+			event := &Event{
+				Params: make([]keyValuePair, 0),
+			}
+
+			if serviceToken := r.Header.Get(a.TokenName); len(serviceToken) > 0 {
+				ctx = context.WithValue(ctx, a.TokenName, serviceToken)
+				event.Service = serviceToken
+			}
+
+			if from := r.Header.Get(fromHeader); len(from) > 0 {
+				ctx = context.WithValue(ctx, fromContextKey, from)
+				event.User = from
+			}
+
+			event.RequestID = r.Header.Get(requestIDHeader)
+			event.Namespace = a.Namespace
+
+			eventString, err := json.Marshal(event)
```

Comment:
`eventString` is []byte not string, so it would be better to call is either just eventBytes

It will read better when you stringify the byte array like so `string(eventBytes)` on line 84